### ISSUE TITLE
Export to ONNX

### ISSUE THREAD

**Issue by mush42** [rwth-i6/returnn#1394]
Hi

Is there a way to convert pretrained `returnn` networks to `ONNX` or at least save the network to `tensorflow's saved model` format?

Best
Musharraf

**Comment by albertz** [comment]
Yes. @Gerstenberger has some experience doing this. See for example https://github.com/rwth-i6/returnn/issues/1236#issuecomment-1339201812:

> yes, i use `compile_tf_graph.py` [from RETURNN `tools/`] for this and then call the tf2onnx tool on the resulting graph.

**Comment by albertz** [comment]
Thanks for the detailed write-up. Maybe you can add this to our documentation or RETURNN wiki?

> Python Script - Expand Me
>
> ...
>
> What we did internally is to write custom graph rewrites, similar to `NativeLstm2` case, and apply the optimizations manually this way. This results in decent speedup; for example using `onnxruntime` custom `MultiHeadAttention` op instead of the `returnn` `SelfAttention` sub-graph.

Maybe you can put those scripts also to `tools` or so?

**Comment by Gerstenberger** [comment]
> Thanks for the detailed write-up. Maybe you can add this to our documentation or RETURNN wiki?

Yes, i can do that

> Maybe you can put those scripts also to tools or so?

I can add the `checkpoint-to-savedmodel` script in the next days.

As for the rewriters, it takes a little bit longer. Parts of it are not even integrated internally yet.
It would be easier to actually integrate it into the `i6_core` repository at some point together with the `tensorflow-to-onnx` job we use so we don't have to manage the same code twice and is independent of the `returnn` version we use.

**Comment by albertz** [comment]
> It would be easier to actually integrate it into the `i6_core` repository at some point together with the `tensorflow-to-onnx` job we use so we don't have to manage the same code twice and is independent of the `returnn` version we use.

It depends always. As soon as there is anything non-trivial, slightly complex logic, even if independent of RETURNN, we would put it in `tools`, and then the job in i6_core would just call this script.

So in principle a user could also run the tool directly and not use Sisyphus / i6_core.

If it's so trivial, basically anyway just one single command to run, which is also just another external script, this of course would not justify an own script in `tools`.

But I think such rewriting thing sounds very much like it should be an own script? And I think RETURNN `tools` is the optimal place for this.
